third_party/setup.py

```diff
@@ -0,0 +1,244 @@
+    with tarfile.open(tempfile, 'r') as archive:
+        for member in archive.getmembers():
+            match = re.match('^[^/]+/', member.name)
+            if match:
+                outname = os.path.join(dir, member.name[match.span(0)[1]:])
+                if member.isdir():
+                    if not os.path.exists(outname):
+                        os.mkdir(outname)
+                elif member.isfile():
+                    with archive.extractfile(member) as infile:
+                        with open(outname, 'wb') as outfile:
+                            outfile.write(infile.read())
+                        if sys.platform != 'win32':
+                            os.chmod(outname, member.mode)
+    os.remove(tempfile)
+
+
+# mozjs
+# see: https://github.com/GoogleChromeLabs/jsvu/tree/main/engines/spidermonkey
+
+mozjs_bin = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'mozjs')
+
+
+def mozjs_determine_platform():
+    is_64bits = sys.maxsize > 2**32
+    if sys.platform == 'linux' or sys.platform == 'linux2':
+        return 'linux-x86_64' if is_64bits else 'linux-i686'
+    if sys.platform == 'darwin':
+        return 'mac'
+    if sys.platform == 'win32':
+        return 'win64' if is_64bits else 'win32'
+    print('Cannot determine platform, assuming \'linux-x86_64\'')
+    return 'linux-x86_64'
+
+
+def mozjs_determine_version(platform):
+    data = fetch_json('https://product-details.mozilla.org/1.0/firefox_history_development_releases.json')
+    latest = ''
+    version = ''
+    for v, t in data.items():
+        if t > latest:
+            latest = t
+            version = v
+    return version
+
+
+def mozjs_download(platform, version):
+    download_zip('https://archive.mozilla.org/pub/firefox/releases/' + version + '/jsshell/jsshell-' + platform + '.zip', mozjs_bin)
+    if sys.platform != 'win32':
+        os.rename(os.path.join(mozjs_bin, 'js'), os.path.join(mozjs_bin, 'mozjs'))
+        os.chmod(os.path.join(mozjs_bin, 'mozjs'), 0o755)
+    else:
+        os.rename(os.path.join(mozjs_bin, 'js.exe'), os.path.join(mozjs_bin, 'mozjs.exe'))
+
+
+def mozjs_is_installed():
+    return os.path.exists(os.path.join(mozjs_bin, 'mozjs.exe' if sys.platform == 'win32' else 'mozjs'))
+
+
+def mozjs_main():
+    print('Setting up mozjs ...')
+    platform = mozjs_determine_platform()
+    print('* Platform: ' + platform)
+    version = mozjs_determine_version(platform)
+    print('* Latest version: ' + version)
+    print('* Downloading to: ' + mozjs_bin)
+    mozjs_download(platform, version)
+    if mozjs_is_installed():
+        print('* Complete')
+    else:
+        print('* Something went wrong :(')
+
+
+# V8
+# see: https://github.com/GoogleChromeLabs/jsvu/tree/main/engines/v8
+
+v8_bin = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'v8')
+
+
+def v8_determine_platform():
+    is_64bits = sys.maxsize > 2**32
+    if sys.platform == 'linux' or sys.platform == 'linux2':
```

Comment:
Should this be `linux32` rather than `linux2`?

Reply:
Apparently the string used to include the major version as returned by `uname -r` prior to Python 3.3. Relevant [python docs](https://docs.python.org/3/library/sys.html#sys.platform) suggest to use `startsWith`, going to change.